### ISSUE TITLE
CQL collections and null keys or values

### ISSUE THREAD

**Issue by jfgosselin** [Impetus/kundera#566]
null values (or keys for map) are valid with CQL collections (set, list and map)  but are discarded by Kundera 

CQLTranslator.java : 

```
private boolean appendMap(StringBuilder builder, Object value)
{
    boolean isPresent = false;
    Map map = ((Map) value);
    isPresent = true;
    builder.append("{");
    for (Object mapKey : map.keySet())
    {
        Object mapValue = map.get(mapKey);
        if (mapKey != null && mapValue != null)       <---- null are ignored 
        {
            appendValue(builder, mapKey.getClass(), mapKey, false);
            builder.append(":");
            appendValue(builder, mapValue.getClass(), mapValue, false);
        }
        builder.append(",");
    }
    if (!map.isEmpty())
    {
        builder.deleteCharAt(builder.length() - 1);
    }
    builder.append("}");
    return isPresent;
}
```

**Comment by mevivs** [comment]
CQL3 column family would have kind of static schema. Even if a column value is not assigned(implicitly "null"), it would be displayed as null only. Similarly if you retrieve it back it would be assigned a null reference. Is there a to convert "null" into a byte[] ? I am also not able to find a reference http://cassandra.apache.org/doc/cql3/CQL.html for NULL.

-Vivek

**Comment by jfgosselin** [comment]
I couldn't find any references but was playing with com.datastax.driver.core.querybuilder.QueryBuilder to validate 

```
Map<String, String> map = new HashMap<>();
map.put("key1", null);
map.put("key2", null);
Insert insert = QueryBuilder.insertInto("foo", "bar");
insert.value("amap", map);
```

This statement will  produce :

INSERT INTO foo.bar(amap) VALUES ({'key2':null,'key1':null});

**Comment by mevivs** [comment]
I agree. But would still wonder what's the point of having a map with null values. At present Cassandra doesn't provide a way to index key or values within collection. Also there is no such "isNull" support. But may be available in future.  

Will discuss and post an update.

-Vivek

**Comment by kkmishra** [comment]
Verified with cassandra-1.2.9 and cassandra-2.0.6 but null is not supported inside collections.

Done Following steps

**create table script**

```
CREATE TABLE blog_posts (  post_id int,  body text,  comments map<int, text>,  liked_by list<int>, PRIMARY KEY(post_id));
```

**Inserting data**

```
cqlsh:KunderaExamples> INSERT INTO "blog_posts"("liked_by","body","comments","post_id") VALUES([111,222],'Working with Cassandra using Kundera',{222:'I am getting NPE on line no. 145',444:'null',111:'What a post!'},1);
cqlsh:KunderaExamples> 
```

**Verifying**

```
cqlsh:KunderaExamples> select * from "blog_posts";

 post_id | body                                 | comments                                                                    | liked_by
---------+--------------------------------------+-----------------------------------------------------------------------------+------------
       1 | Working with Cassandra using Kundera | {111: 'What a post!', 222: 'I am getting NPE on line no. 145', 444: 'null'} | [111, 222]

(1 rows)
```

**While inserting null for collection I am getting following error.**

```
cqlsh:KunderaExamples> INSERT INTO "blog_posts"("liked_by","body","comments","post_id") VALUES([111,222],'Working with Cassandra using Kundera',{222:'I am getting NPE on line no. 145',444:null,111:'What a post!'},1);
Bad Request: null is not supported inside collections
```

Thanks
Kuldeep

**Comment by jfgosselin** [comment]
You're right, I did a test with the java driver and I get :
Exception in thread "main" com.datastax.driver.core.exceptions.InvalidQueryException: null is not supported inside collections

The problem that I saw is that CQLTranslator will send e.g. {,,,,} if a map has only null values and it will throw an InvalidRequestException

Caused by: InvalidRequestException(why:line 1:105 no viable alternative at input '}') .
ERROR 2014-04-03 10:47:51,930 [Thread-5-MainRunner] com.impetus.client.cassandra.CassandraClientBase - Error while persisting record. Caused by: .
javax.persistence.PersistenceException: InvalidRequestException(why:line 1:105 no viable alternative at input '}')

**Comment by mevivs** [comment]
So are we good to close on this? As Cassandra doesn't allow map with null values.

-Vivek

**Comment by jfgosselin** [comment]
Yes, but the question is should Kundera send e.g. {'key1':null, 'key2':null} instead of {,} and let Cassandra return the InvalidQueryException.

**Comment by chhavigangwal** [comment]
Fixed in Kundera-2.11.1
